jax/_src/numpy/lax_numpy.py

```diff
@@ -2467,21 +2467,15 @@ def diag_indices_from(arr):
 @partial(jit, static_argnames=('offset', 'axis1', 'axis2'))
 def diagonal(a, offset=0, axis1: int = 0, axis2: int = 1):
   _check_arraylike("diagonal", a)
-  a_shape = shape(a)
-  a_ndims = len(a_shape)
   offset = core.concrete_or_error(operator.index, offset, "'offset' argument of jnp.diagonal()")
 
-  # Move the two dimensions to the end.
-  axis1 = _canonicalize_axis(axis1, a_ndims)
-  axis2 = _canonicalize_axis(axis2, a_ndims)
-  perm = [i for i in range(a_ndims) if i != axis1 and i != axis2]
-  perm = perm + [axis1, axis2]
-  a = lax.transpose(a, perm)
+  a = moveaxis(a, (axis1, axis2), (-2, -1))
 
+  a_shape = shape(a)
   diag_size = _max(0, _min(a_shape[axis1] + _min(offset, 0),
                            a_shape[axis2] - _max(offset, 0)))
-  i = arange(diag_size)
```

Comment:
I would stick with arange here. It will generate an `iota` HLO op, rather than a constant containing the same values. This is preferable because `iota` is more compact to represent.

Reply:
> It will generate an iota HLO op, rather than a constant containing the same values.

Makes sense, sorry I didn't catch this since `arange(_abs(offset), _abs(offset) + diag_size)` currently always generates a constant so I missed it when looking at the HLO.

:+1: I reverted the change and updated the PR to also use `jnp.moveaxis` in `jnp.trace`.